### ISSUE TITLE
random matrix over Galois field

### ISSUE THREAD

**Issue by naddington** [Macaulay2/M2#978]
In 1.14, the following code returns the same answer every time, both on my machine and another, and on habanero:

k = GF(81)
random(k^2, k^2)

setRandomSeed has no effect.  I get the same behavior with differently sized matrices, even 1x1.  But if I use random(k) then I get different answers every time, or if I replace GF(81) with ZZ/103.

Thanks,
Nick Addington

**Comment by naddington** [comment]
Interestingly, the problem occurs in GF(103) but not ZZ/103.

**Comment by DanGrayson** [comment]
On my Mac it works:
```
**Macaulay2, version 1.14
--loading configuration for package "FourTiTwo" from file /Users/dan/Library/Application Support/Macaulay2/init-FourTiTwo.m2
--loading configuration for package "Topcom" from file /Users/dan/Library/Application Support/Macaulay2/init-Topcom.m2
with packages: ConwayPolynomials, Elimination, IntegralClosure, InverseSystems, LLLBases,
               PrimaryDecomposition, ReesAlgebra, TangentCone, Truncations

i1 : k = GF(81)

o1 = k

o1 : GaloisField

i2 : netList for i to 10 list random(k^2, k^2)

     +------------------------+
o2 = || a3+a2+a a3-a2  |      |
     || -a2-a   a2+a-1 |      |
     +------------------------+
     || -a3+a2-a-1 a3-a+1  |  |
     || a2+a-1     a3+a2-a |  |
     +------------------------+
     || -a3-a2+a-1 a3-1  |    |
     || a3+a2+1    -a2-a |    |
     +------------------------+
     || a3+a2-a+1 -a3+a2-1 |  |
     || -a3+a     -a3-a2   |  |
     +------------------------+
     || -a3+a2+1 -a2+a     |  |
     || -a3      a3+a2+a+1 |  |
     +------------------------+
     || a3-a a3+a+1  |        |
     || 0    -a2+a+1 |        |
     +------------------------+
     || a3+a+1 a3+a2-a |      |
     || a3-a+1 a3+a2+a |      |
     +------------------------+
     || a3+a2+a+1 a3+a2+a-1  ||
     || -1        -a3-a2-a+1 ||
     +------------------------+
     || a3-a2+a+1 a3-a2-1    ||
     || a3+a      -a3+a2-a-1 ||
     +------------------------+
     || -a3+a2+1  -a2-a |     |
     || a3+a2-a+1 a+1   |     |
     +------------------------+
     || -a3-a2+1 -a3+1   |    |
     || -a+1     -a2-a+1 |    |
     +------------------------+
```

**Comment by naddington** [comment]
If you run both lines over and over (not just the second line) you should see the behavior. Your o2 is what I always get the first time.

**Comment by DanGrayson** [comment]
Oh!  Then this result shows that the wrong random number generator is being used here:
```
Macaulay2, version 1.14
--loading configuration for package "FourTiTwo" from file /Users/dan/Library/Application Support/Macaulay2/init-FourTiTwo.m2
--loading configuration for package "Topcom" from file /Users/dan/Library/Application Support/Macaulay2/init-Topcom.m2
with packages: ConwayPolynomials, Elimination, IntegralClosure, InverseSystems, LLLBases,
               PrimaryDecomposition, ReesAlgebra, TangentCone, Truncations

i1 : k = GF(81)

o1 = k

o1 : GaloisField

i2 : netList for i to 10 list (setRandomSeed 0; random(k^2, k^2))

     +------------------------+
o2 = || a3+a2+a a3-a2  |      |
     || -a2-a   a2+a-1 |      |
     +------------------------+
     || -a3+a2-a-1 a3-a+1  |  |
     || a2+a-1     a3+a2-a |  |
     +------------------------+
     || -a3-a2+a-1 a3-1  |    |
     || a3+a2+1    -a2-a |    |
     +------------------------+
     || a3+a2-a+1 -a3+a2-1 |  |
     || -a3+a     -a3-a2   |  |
     +------------------------+
     || -a3+a2+1 -a2+a     |  |
     || -a3      a3+a2+a+1 |  |
     +------------------------+
     || a3-a a3+a+1  |        |
     || 0    -a2+a+1 |        |
     +------------------------+
     || a3+a+1 a3+a2-a |      |
     || a3-a+1 a3+a2+a |      |
     +------------------------+
     || a3+a2+a+1 a3+a2+a-1  ||
     || -1        -a3-a2-a+1 ||
     +------------------------+
     || a3-a2+a+1 a3-a2-1    ||
     || a3+a      -a3+a2-a-1 ||
     +------------------------+
     || -a3+a2+1  -a2-a |     |
     || a3+a2-a+1 a+1   |     |
     +------------------------+
     || -a3-a2+1 -a3+1   |    |
     || -a+1     -a2-a+1 |    |
     +------------------------+
```

**Comment by DanGrayson** [comment]
Mike, I think you can fix this easily.

**Comment by mikestillman** [comment]
Fixed in PR #2493.  Finally!  Sorry for not fixing this earlier.  I had to change the random number generation function for Galois fields, as the one that was being used seemed to reset back to its original state each time the same ring was created.

**Comment by naddington** [comment]
Thanks for letting me know!
